File: src/lib/koboEdit.ts
const EU_HOSTNAME = "eu.kobotoolbox.org";

export function resolveKoboEditToken(
  server: string,
  tokens: { global: string; eu: string }
): string {
  try {
    const host = new URL(server).hostname;
    return host === EU_HOSTNAME ? tokens.eu : tokens.global;
  } catch {
    return tokens.global;
  }
}

/**
 * Queries the Kobo data API to find the numeric `_id` for a submission
 * identified by its `_uuid`. Returns null if not found or on error.
 */
export async function resolveSubmissionId(
  server: string,
  uid: string,
  uuid: string,
  token: string
): Promise<number | null> {
  try {
    const query = JSON.stringify({ _uuid: uuid });
    const fields = JSON.stringify(["_id"]);
    const url = `${server}/api/v2/assets/${uid}/data.json?query=${encodeURIComponent(query)}&fields=${encodeURIComponent(fields)}`;
    const res = await fetch(url, {
      headers: { Authorization: `Token ${token}` },
    });
    if (!res.ok) {
      console.error(`[edit] resolveSubmissionId failed: HTTP ${res.status}`);
      return null;
    }
    const data = await res.json<{ results?: Array<{ _id?: number }> }>();
    return data.results?.[0]?._id ?? null;
  } catch (err) {
    console.error("[edit] resolveSubmissionId error:", err);
    return null;
  }
}

/**
 * Patches field values back onto an existing Kobo submission using the
 * bulk-edit endpoint. The `data` map uses question xpaths as keys.
 * No binary content is sent — only JSON field values.
 */
export async function editSubmission(
  server: string,
  uid: string,
  id: number,
  data: Record<string, string>,
  token: string
): Promise<{ ok: boolean; httpStatus: number; error?: string }> {
  try {
    const url = `${server}/api/v2/assets/${uid}/data/bulk/`;
    const payload = { submission_ids: [id], data };
    const res = await fetch(url, {
      method: "PATCH",
      headers: {
        Authorization: `Token ${token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ payload: JSON.stringify(payload) }),
    });
    if (!res.ok) {
      const text = await res.text().catch(() => "");
      console.error(`[edit] editSubmission failed: HTTP ${res.status} — ${text.slice(0, 200)}`);
      return { ok: false, httpStatus: res.status, error: text.slice(0, 500) };
    }
    return { ok: true, httpStatus: res.status };
  } catch (err) {
    console.error("[edit] editSubmission error:", err);
    return { ok: false, httpStatus: 0, error: String(err) };
  }
}

File: app/lib/kobo_edit.py
"""
Kobo submission edit-back helpers.
Port of src/lib/koboEdit.ts.
"""

import json
import logging

import httpx

logger = logging.getLogger(__name__)

EU_HOSTNAME = 'eu.kobotoolbox.org'


def resolve_kobo_edit_token(server: str, tokens: dict[str, str]) -> str:
    try:
        from urllib.parse import urlparse
        host = urlparse(server).hostname or ''
        return tokens['eu'] if host == EU_HOSTNAME else tokens['global']
    except Exception:
        return tokens['global']


def resolve_submission_id(server: str, uid: str, uuid: str, token: str) -> int | None:
    """Queries Kobo to find the numeric _id for a submission identified by _uuid."""
    try:
        query = json.dumps({'_uuid': uuid})
        fields = json.dumps(['_id'])
        url = f'{server}/api/v2/assets/{uid}/data.json'
        resp = httpx.get(
            url,
            params={'query': query, 'fields': fields},
            headers={'Authorization': f'Token {token}'},
            timeout=15,
        )
        if not resp.is_success:
            logger.error('[edit] resolveSubmissionId failed: HTTP %s', resp.status_code)
            return None
        data = resp.json()
        results = data.get('results') or []
        return results[0].get('_id') if results else None
    except Exception as exc:
        logger.error('[edit] resolveSubmissionId error: %s', exc)
        return None


def edit_submission(
    server: str, uid: str, submission_id: int, data: dict[str, str], token: str
) -> dict:
    """Patches field values onto an existing Kobo submission via the bulk-edit endpoint."""
    try:
        url = f'{server}/api/v2/assets/{uid}/data/bulk/'
        payload_inner = json.dumps({'submission_ids': [submission_id], 'data': data})
        resp = httpx.patch(
            url,
            json={'payload': payload_inner},
            headers={'Authorization': f'Token {token}'},
            timeout=15,
        )
        if not resp.is_success:
            text = resp.text[:500]
            logger.error('[edit] editSubmission failed: HTTP %s — %s', resp.status_code, text[:200])
            return {'ok': False, 'http_status': resp.status_code, 'error': text}
        return {'ok': True, 'http_status': resp.status_code}
    except Exception as exc:
        logger.error('[edit] editSubmission error: %s', exc)
        return {'ok': False, 'http_status': 0, 'error': str(exc)}


def update_validation_status(
    server: str, uid: str, submission_id: int, status: str, token: str
) -> dict:
    """Sets the validation status of a Kobo submission."""
    try:
        url = f'{server}/api/v2/assets/{uid}/data/{submission_id}/validation_status/'
        resp = httpx.patch(
            url,
            data={'validation_status.uid': status},
            headers={'Authorization': f'Token {token}'},
            timeout=15,
        )
        if not resp.is_success:
            text = resp.text[:500]
            logger.error('[validate] updateValidationStatus failed: HTTP %s — %s', resp.status_code, text[:200])
            return {'ok': False, 'http_status': resp.status_code, 'error': text}
        return {'ok': True, 'http_status': resp.status_code}
    except Exception as exc:
        logger.error('[validate] updateValidationStatus error: %s', exc)
        return {'ok': False, 'http_status': 0, 'error': str(exc)}
